### ISSUE TITLE
Normalise language list in filters

### ISSUE THREAD

**Issue by sarahmonster** [opentexts/opentexts#87]
I was originally filing this as part of #86 but I realise that it's a slightly different problem, so probably best split into a separate issue.

@brizee mentioned during a conversation we had a week or two ago that there's similarly messy data in the language. For example:

- Not Specified, Undetermined, No Linguistic Content (could we combine these, maybe, and give them a more user-friendly label?)
- eng → should be compiled into English
- should we simplify the labels for older variants of a language (Greek, English, German, French etc)? Including the dates here makes these labels long.
- we should probably remove "Other" from language labels (it's quite literally othering!)
- consider grouping all languages with less than twenty results under a single label (very mixed feelings about this)

Here's an example of a language filter that I get if I search "books":

English (2,450,412)
German (221,542)
French (207,016)
Undetermined (78,050)
Spanish (60,241)
Italian (47,065)
Latin (27,905)
Russian (15,113)
Dutch (13,674)
Portuguese (9,743)
Danish (8,018)
No linguistic content (7,687)
Swedish (4,114)
Korean (3,872)
Yiddish (3,217)
Polish (3,149)
Czech (2,778)
Hungarian (2,238)
Greek, Ancient (to 1453) (2,137)
Catalan (1,217)
Persian (1,182)
Greek, Modern (1453-) (1,175)
Not specified (1,162)
Chinese (1,126)
Arabic (897)
Norwegian (759)
Japanese (738)
Multiple languages (698)
Armenian (653)
Hebrew (586)
Ukrainian (562)
Syriac (521)
Icelandic (465)
Croatian (464)
Serbian (448)
Welsh (408)
Turkish, Ottoman (384)
Finnish (336)
Pushto (325)
French, Middle (ca. 1300-1600) (243)
Slovak (216)
Tagalog (198)
Greek (186)
Irish (177)
Afrikaans (134)
Romanian (134)
Romance (Other) (133)
German, Middle High (ca. 1050-1500) (110)
Slovenian (80)
Bulgarian (79)
English, Old (ca. 450-1100) (70)
Turkish (68)
French, Old (ca. 842-1300) (67)
Germanic (Other) (60)
Urdu (58)
Austronesian (Other) (53)
Latvian (52)
Sanskrit (49)
Lithuanian (48)
Frisian (47)
eng (47)
Scottish Gaelic (45)
Syriac, Modern (45)
Church Slavic (39)
Hindi (37)
Tahitian (31)
Occitan (post-1500) (30)
Slavic (Other) (28)
Hawaiian (26)
Vietnamese (25)
Marathi (22)
Indonesian (21)
Estonian (20)
Makasar (20)
Pampanga (20)
Dutch, Middle (ca. 1050-1350) (19)
English, Middle (1100-1500) (19)
Nahuatl (19)
Pahlavi (17)
Bugis (16)
Creoles and Pidgins, French-based (Other) (16)
Miscellaneous languages (16)
Pali (16)
Malay (15)
Malayalam (15)
Thai (15)
Bantu (Other) (13)
Bengali (13)
Cebuano (13)
Khmer (13)
Norwegian (Bokmål) (13)
Esperanto (11)
German, Old High (ca. 750-1050) (11)
Iloko (11)
Sino-Tibetan (Other) (11)
Sorbian (Other) (11)
Tamil (11)
Albanian (10)
Gujarati (10)
Niger-Kordofanian (Other) (10)

Definitely can get a bit overwhelming! Adding the ability to search within filters should help here, but is there more we can do to make this list easier to parse?

**Comment by stuartlewis** [comment]
I've tieded this up a little.  It turned out that there was some rogue data that had crept in to the solr index which I have now removed. They were easy to identify and therefore remove, as they had no associated library name. For my own future reference, the command to remove these was: 

```bash
bin/post -c otw -d '<delete><query>-organisation:["" TO *]</query></delete>'
```

This has removed the non-translated languages such as eng / fre / ger, etc.

The formal list of languages we work from is [from the Library of Congress](https://www.loc.gov/marc/languages/language_code.html). Unfortunatly it is very verbose, but I guess at least it means it is 'accurate' if someone is looking for materials in a language that doesn't have so much digitised content available. This will also become an even bigger problem in the future when we try to transalte or internationalise the site, as we will then need translations of each of these (in which case going back and storing codes rather than language names in English might be a better approach anyway)?

A lot of this comes down to the real challenge of aggregating metadata create by huyndreds of libraries.  We all work to the same standards... but interpret them differently! 👎

**Comment by brizee** [comment]
It might be verbose but I love a good standard! :smile: maybe we could form some sensible sub-categories? Time periods feels like a natural way of grouping them perhaps for people who are _discovering_ more than searching? Though ultimately a search box is probably the winner as long it is standardised.

**Comment by stuartlewis** [comment]
Yes - I definitely want to do a year filter if we can (and in solr it should be possible to graduate the choices (e.g. first into centuries, then decades, then years, etc).  However I need to re-parse and re-ingest all the data again I think, and reconfigure solr to know that year is a date field.  There is a lot of unprecise data in that field, for exmaple libraries put things like 18uu if they don't know when in the 1800's an item was published.

Howevert this won't be wasted effort, as I can also reformat all the data, for example where I've added more precise url fields for the different file types.

**Comment by brizee** [comment]
oh fun 😁 - maybe we need a searchable date as well as a presentable date for things like 18uu?

Though in my case I was thinking for languages - rather than opening the Languages and seeing every language. Put together a crude mock:
![image](https://user-images.githubusercontent.com/188324/91731793-39301000-eb9f-11ea-8612-dfa100feb86a.png)

Just an idea, don't worry I'm not precious about them but might make exploring a bit easier?

**Comment by stuartlewis** [comment]
Wow - that would be pretty neat!  Any idea how to do this (would you go for [hierarchical faceting in solr](https://cwiki.apache.org/confluence/display/solr/HierarchicalFaceting) or do it in the application layer?

Yup, with years we could definitely make use of searchable and display years.   I might see how easy it will be to just force everything into yyyy anyway and see what we lose doing that, as I think things like 18uu probably aren't understood by many people anyway.

**Comment by brizee** [comment]
Technically this isn't changing anything about functionality - just how it's presented - if you wanted to Select all middle languages then I'd probably do that in application layer though that's as much about my lack of familiarity with Solr / the data import process.

**Comment by brizee** [comment]
Just making title on this a bit more accurate, since normalisation isn't the problem here.

**Comment by lb42** [comment]
If you plan to define language codes hierarchically, don't forget to cater for regional variation as well as historical period. So for example distinguishing Belgian, Canadian, and metropolitan French; and don't get me started on Swiss languages.
